src/credential_provider.rs

```diff
@@ -0,0 +1,473 @@
+use crate::{MomentoError, MomentoResult};
+use base64::Engine;
+use jsonwebtoken::{decode, Algorithm, DecodingKey, Validation};
+use serde::{Deserialize, Serialize};
+use std::env;
+
+#[derive(Debug, Serialize, Deserialize)]
+struct JwtClaims {
+    sub: String,
+    c: Option<String>,
+    cp: Option<String>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct V1Token {
+    pub api_key: String,
+    pub endpoint: String,
+}
+
+const CACHE_ENDPOINT_PREFIX: &str = "cache.";
+const CONTROL_ENDPOINT_PREFIX: &str = "control.";
+
+#[derive(Debug, Clone)]
+pub struct CredentialProvider {
+    pub auth_token: String,
+    pub control_endpoint: String,
+    pub cache_endpoint: String,
+}
+
+#[derive(Debug)]
+pub struct CredentialProviderBuilder<'a> {
+    auth_token_string: Option<String>,
+    auth_token_env_var: Option<&'a str>,
+    cache_endpoint_override: Option<String>,
+    control_endpoint_override: Option<String>,
+}
+
+impl<'a> CredentialProviderBuilder<'a> {
+    pub fn new_from_environment_variable(env_var_name: &'a str) -> Self {
+        CredentialProviderBuilder {
+            auth_token_env_var: Some(env_var_name),
+            auth_token_string: None,
+            cache_endpoint_override: None,
+            control_endpoint_override: None,
+        }
+    }
+
+    pub fn new_from_string(auth_token: String) -> Self {
+        CredentialProviderBuilder {
+            auth_token_env_var: None,
+            auth_token_string: Some(auth_token),
+            cache_endpoint_override: None,
+            control_endpoint_override: None,
+        }
+    }
+
+    pub fn with_cache_endpoint(mut self, cache_endpoint_override: Option<String>) -> Self {
```

Comment:
why do you need `Option` in these signatures?

Reply:
In cli code we sometimes have an endpoint override (with `-e` flag) - so that comes in as `Option<String>`. It's easier to do the builder chaining if these `with...` take an option... but I think that's not actually the right solution.

Reply:
Ok I figured out what I was doing wrong in the CLI code and fixed this in befa730.